src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java

```diff
@@ -157,9 +163,28 @@
         p.setDefinition(def);
         WorkflowRun b = r.assertBuildStatusSuccess(p.scheduleBuild2(0));
         r.assertLogNotContains("Cloning the remote Git repository", b);
+        r.assertLogNotContains("Retrying after 10 seconds", b);
         r.assertLogContains("Obtained flow.groovy from git " + sampleRepo, b);
         r.assertLogContains("version one", b);
     }
+    
+    @Issue("JENKINS-39194")
+    @Test public void retry() throws Exception {
+        // We use an un-initialized repo here to test retry
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        GitStep step = new GitStep(invalidRepo.toString());
+        CpsScmFlowDefinition def = new CpsScmFlowDefinition(step.createSCM(), "flow.groovy");
+        def.setLightweight(false);
+        p.setDefinition(def);
+        // Initially check for the global Jenkins retry count
+        assertEquals(r.jenkins.getScmCheckoutRetryCount(), def.getScmCheckoutRetryCount());
+        // Set the jenkins global retry count to 3 and ensure that they are still equal
+        r.jenkins.setScmCheckoutRetryCount(1);
+        assertEquals(r.jenkins.getScmCheckoutRetryCount(), def.getScmCheckoutRetryCount());
+        WorkflowRun b = r.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0));
+        r.assertLogContains("Could not read from remote repository", b);
+        r.assertLogContains("Retrying after 10 seconds", b);
```

Comment:
Could I suppose use `waitForMessage`, initialize the repository, then assert then the build completes.